src/CSET/operators/convection.py

```diff
@@ -0,0 +1,137 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""A module containing different diagnostics for convection.
+
+The diagnostics are calculated from output from the Unified Model, although
+precalculated values in the required input form may also be used.
+
+"""
+
+import copy
+
+import numpy as np
+
+
+def cape_ratio(SBCAPE, MUCAPE, MUCIN, MUCIN_thresh=-75.0):
+    r"""Ratio of two fields, one filtered to allow physical values to be output.
+
+    Parameters
+    ----------
+    SBCAPE: cube
+        Surface-based convective available potential energy as calculated by the
+        model.
+        Stash: m01s20i114
+    MUCAPE: cube
+        Most-unstable convective available potential energy as calculated by the
+        model.
+        Stash: m01s20i112
+    MUCIN: cube
+        Most-unstable convective inhibition associated with the most-unstable
+        ascent as calculated by the model.
+        Stash: m01s20i113
+    MUCIN_thresh: float, optional, default is -75. J/kg.
+        Threshold to filter the MUCAPE by values are realistically realisable.
+
+    Returns
+    -------
+    cube
+
+    Notes
+    -----
+    This diagnostic is based on Clark et al. (2012) [1]_. It is based around the idea
+    that for elevated convection the convective instability is not based at the
+    surface. This utilises two flavours of CAPE: the surface-based CAPE (SBCAPE)
+    and the most-unstable CAPE (MUCAPE). The MUCAPE is filtered by the MUCIN
+    associated with that parcel's ascent to ensure that any CAPE can at least
+    theoretically be released. The default value is set at -75 J/kg but it can
+    be changes depending on location and users requirements.
+
+    .. math:: 1 - (\frac{SBCAPE}{MUCAPE})
+
+    The ratio is defined in this way such that if SBCAPE=MUCAPE the ratio will
+    equal 1. If the ratio was reversed when MUCAPE exists and SBCAPE is zero the
+    ratio would be undefined.
+
+    The diagnostic varies smoothly between zero and unity. A value of 0 implies
+    an environment is suitable for surface-based convection. A value of 1
+    implies an environment is suitable for elevated convection. Values between
+    imply transition convection with values closer to one imply elevated
+    convection is more likely and values closer to zero implying that
+    surface-based convection is more likely.
+
+    Further details about this diagnostic for elevated convection identification
+    can be found in Flack et al. (2023) [2]_.
+
+    Expected applicability ranges: Convective-scale models will be noisier than
+    parametrized models as they are more responsive to the convection, and thus
+    it may be more sensible to view as a larger spatial average rather than
+    grid point.
```

Comment:
```suggestion
    Expected applicability ranges: Convective-scale models will be noisier than
    parametrized models as they are more responsive to the convection, and thus
    it may be more sensible to view as a larger spatial average rather than
    on the native resolution.
```